### ISSUE TITLE
Proposal: change API for modifications

### ISSUE THREAD

**Issue by odow** [JuliaOpt/MathOptInterface.jl#360]
To date, I'm not sure if anyone has dug into the modification API yet. I was about to start writing tests (https://github.com/JuliaOpt/MathOptInterface.jl/pull/354) and had some thoughts.

---

In addition to some [function modifications](http://www.juliaopt.org/MathOptInterface.jl/latest/apireference.html#MathOptInterface.ScalarConstantChange), we currently have two [modifyconstraint!](http://www.juliaopt.org/MathOptInterface.jl/latest/apireference.html#MathOptInterface.modifyconstraint!) methods:

`modifyconstraint!(m, c::ConstraintIndex{F,S}, new_set::S)`
`modifyconstraint!(m, c::ConstraintIndex{F,S}, new_function::F)`

Considering the rest of the API, doesn't it make more sense to make these `set!` methods?

```julia
set!(m::ModelLike, ::ConstraintFunction, c::ConstraintIndex{F,S}, new_function::F)
set!(m::ModelLike, ::ConstraintSet, c::ConstraintIndex{F,S}, new_set::S)
```

---

Furthermore, instead of `modifyconstraint!` and `modifyobjective!`, we could have
```julia
canmodify(m, ::ConstraintFunction, c::Type{ConstraintIndex{F,S}}, change::Type{AbstractFunctionModification})
modify!(m, ::ConstraintFunction, c::ConstraintIndex{F,S}, change::AbstractFunctionModification)

canmodify(m, ::ObjectiveFunction, change::Type{AbstractFunctionModification})
modify!(m, ::ObjectiveFunction, change::AbstractFunctionModification)
```

Edit: we could also change the `transformconstraint` to 
```julia
canmodify(m, ::ConstraintSet, c::Type{ConstraintIndex{F,S1}}, ::Type{S2})
c2 = modify!(m, ::ConstraintSet, c::ConstraintIndex{F,S1}, new_set::S2)
```

**Comment by blegat** [comment]
I am in favor of these changes.

For `transform` -> `modify`, I remember we named it transform to make it different than `modify` with a set of the same type but if modifying the set is moved to `set!` then we could add changing the set with a new set type to `modify`. However, `transform` need to return a new constraint index while `modify` does not change the constraint index so it is probably better to keep then separate functions.

**Comment by mlubin** [comment]
Agreed with @blegat. The change to `set!` and `modify!` looks good. I disagree with merging `transformconstraint` into `modify!` because the function has a different meaning that's inconsistent with what happens when the sets are of the same type.

**Comment by odow** [comment]
The majority of this was addressed by #388.

Closing in favor of #398.
